### ISSUE TITLE
Bug: "Single run" not resolving types / resolving incorrect libs

### ISSUE THREAD

**Issue by milesj** [typescript-eslint/typescript-eslint#5136]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```typescript
I'm not sure what the conditions are to reproduce this yet.
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  parserOptions: {
    // This path changes based on which folder the eslint binary is running in
    project: ['path/to/tsconfig.json'],
    tsconfigRootDir: __dirname,
    allowAutomaticSingleRunInference: true,
  },
  extends: ['airbnb', 'import', 'react', 'react-hooks', 'jest', '@typescript-eslint'],
  reportUnusedDisableDirectives: true,
  rules: {
    // Pretty much all the rules provided by the plugins/configs
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "baseUrl": ".",
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "jsx": "react-jsx",
    "isolatedModules": true,
    "lib": [
      "dom",
      "esnext"
    ],
    "module": "CommonJS",
    "moduleResolution": "node",
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "ES2020",
  },
  "include": [
    "**/*"
  ]
}
```

### Expected Result

When using `allowAutomaticSingleRunInference` or `TSESTREE_SINGLE_RUN=true`, types and lints resolve/pass correctly, in the same manner as if these settings were turned off.

### Actual Result

When using `allowAutomaticSingleRunInference` or `TSESTREE_SINGLE_RUN=true`, we've noticed that types (mainly `@types`) spuriously fail to resolve. For example, `useMemo` from `import { useMemo } from 'react'` resolves to `any`, causing our `@typescript-eslint/no-unsafe-*` to fail.

When looking at the debug logs, we noticed that the resolved `lib`s are _not what we configured_ (the `esnext.full` is no where to be found in our repo). Once this incorrect lib is used, types start failing. Here's an excerpt from the logs:

```
eslint:cli-engine Lint <path>/config/atomic.ts +8ms
typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(1) {
	'<path>/tsconfig.json'
} +7ms
typescript-eslint:typescript-estree:useProvidedProgram Retrieving ast for <path>/config/atomic.ts from provided program instance(s) +7ms
typescript-eslint:parser:parser Resolved libs from program: [ 'dom', 'esnext' ] +8ms

eslint:cli-engine Lint <path>/config/device.ts +5ms
typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(1) {
	'<path>/tsconfig.json'
} +5ms
typescript-eslint:typescript-estree:createIsolatedProgram Getting isolated program in TS mode for: <path>/config/device.ts +41ms
typescript-eslint:parser:parser Resolved libs from program: [ 'esnext.full' ] +6ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +92ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +1ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +0ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +0ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +1ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +0ms
typescript-eslint:eslint-plugin:utils:types Found an "error" any type +0ms

eslint:cli-engine Lint <path>/config/keys.ts +17ms
typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(1) {
	'<path>/tsconfig.json'
} +18ms
typescript-eslint:typescript-estree:useProvidedProgram Retrieving ast for <path>/config/keys.ts from provided program instance(s) +23ms
typescript-eslint:parser:parser Resolved libs from program: [ 'dom', 'esnext' ] +16ms
```

Notice that `config/device.ts` creates a new isolated program with `esnext.full`, instead of reusing an existing program like `config/atomic.ts` and `config/keys.ts` does, which is `dom, esnext`. All of these files are in the same program, even the same folder.

HOWEVER! When I run the linter on `config/device.ts` individually, the linting passes correctly.

```
eslint:cli-engine Lint <path>/config/device.ts +0ms
typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(1) {
	'<path>/tsconfig.json'
} +0ms
typescript-eslint:typescript-estree:useProvidedProgram Retrieving ast for <path>/config/device.ts from provided program instance(s) +0ms
typescript-eslint:typescript-estree:parser Detected single-run/CLI usage, creating Program once ahead of time for project: <path>/tsconfig.json +0ms
typescript-eslint:parser:parser Resolved libs from program: [ 'dom', 'esnext' ] +0ms
eslint:cli-engine Linting complete in: 3306ms +2s
```

This leads me to believe that the parser is losing the program? Or handling paths/detection wrong? I did some digging into `node_modules` but was unable to find the root cause.

### Additional Info

Our codebase has over 10,000 files and this is consistently running out of memory, so this is me digging into alternatives/why it's happening.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.26.0` |
| `@typescript-eslint/parser`            | `5.26.0` |
| `@typescript-eslint/scope-manager`     | `5.26.0` |
| `@typescript-eslint/typescript-estree` | `5.26.0` |
| `@typescript-eslint/type-utils`        | `5.26.0` |
| `@typescript-eslint/utils`             | `5.26.0` |
| `TypeScript`                           | `4.6.3` |
| `ESLint`                               | `8.16.0` |
| `node`                                 | `16.13.0` |

**Comment by milesj** [comment]
@bradzacher Thanks for the quick response.

Do you have any guidance of where in this repo/package I should debug further? I'll try some more and see if there's a root cause.

**Comment by bradzacher** [comment]
Do you use different tsconfigs for your build and lint? If so then likely there is a difference between the two.

Are there multiple tsconfigs? Is the problem only present in one or all of the projects?

I can keep trying to provide abstract solutions - but it's going to be hard to do much more than that without more concrete info.

If you could try reproducing this in an isolated repo then that would be aces!
Personally I've found with these sorts of problems that attempting to rebuild the setup anew in a fresh repo very quickly reveals the problem!
But I understand that might be hard to do.

**Comment by milesj** [comment]
> Do you use different tsconfigs for your build and lint? 

It's the same config for typechecking and linting.

> Are there multiple tsconfigs? Is the problem only present in one or all of the projects?

Each project has its own (project refs), so the problem applies to all of them.

Yah, this is a multi-year old repo with tons of projects and tens of thousands of files, so trying to replicate it is almost impossible.

But for context, we do the `tsconfig.eslint.json` trick for editors, since we use project refs. When running on the CLI, we use the closest parent `tsconfig.json` (since its usually the project ref itself), which is the config above, and this works EXCEPT when trying to use this "single run" mode.

It then breaks trying to resolve `@types` node modules. I'm trying to use single run mode to avoid the program spin up cost.

**Comment by bradzacher** [comment]
Could you clone the folder structure and tsconfigs, add a handful of dummy files and try to repro specifically the `@types` bug?

If we can repro that we can probably help solve that and that might be your problem sorted.

**Comment by milesj** [comment]
I'll see what I can do tomorrow.

**Comment by bradzacher** [comment]
not yet sorry! family stuff and business trip have gotten in the way.
it's definitely top of my list of things to do!

**Comment by bradzacher** [comment]
https://github.com/milesj/typescript-eslint-single-run/pull/1

The `eslint-plugin-import` lint rules are causing issues.
[As mentioned in our FAQ](https://typescript-eslint.io/docs/linting/troubleshooting#eslint-plugin-import) - there are a few lint rules that you really don't want/need because they're covered by TS.
The rules are pretty bad for perf as well because they do their own out-of-band parsing in order to function.

I'm not sure *why* exactly their out-of-band parsing is causing issues with the single-run mode. When they do their out-of-band parsing [I made it so that they specifically do not pass in `parserOptions.project`](https://github.com/import-js/eslint-plugin-import/blob/376747914b47fbdcf99212b9e9bd4d5e09825385/utils/parse.js#L71-L76) - meaning it creates its own isolated program for the file (driven by [this function](https://github.com/typescript-eslint/typescript-eslint/blob/efc147b04dce52ab17415b6a4ae4076b944b9036/packages/typescript-estree/src/create-program/createIsolatedProgram.ts)).

The exact cause will require more investigation - but at the very least turning off those lint rules should fix your problems.

**Comment by milesj** [comment]
@bradzacher That's wild that `import` rules affect it in that way. Thanks for taking a look at it, I'll make these changes upstream and report back.

**Comment by milesj** [comment]
@bradzacher Still seeing some weirdness when actually pulling it into our repo. I disabled all of these rules:

```
    'import/default': 'off',
    'import/named': 'off',
    'import/namespace': 'off',
    'import/no-cycle': 'off',
    'import/no-named-as-default': 'off',
    'import/no-named-as-default-member': 'off',
    'import/no-deprecated': 'off',
    'import/no-unused-modules': 'off',
```

And also enabled these in `parserOptions`:

```
  allowAutomaticSingleRunInference: false,
  EXPERIMENTAL_useSourceOfProjectReferenceRedirect: true,
```

And linting is still failing for `@typescript-eslint/no-unsafe-*` rules, but in a very unexpected way. For example, this error:

```
error: Unsafe assignment of an `any` value (@typescript-eslint/no-unsafe-assignment) at src/normalizeLocale.ts:2:9:
  1 | export function normalizeLocale(locale: string): string {
> 2 |   const languageParts = locale.split('-');
    |         ^
  3 |
```

Is failing on this file... THIS IS THE ENTIRE FILE. No imports, no other types, etc.

```js
export function normalizeLocale(locale: string): string {
  const languageParts = locale.split('-');

  // ...

  return languageCode;
}
```

I keep trying different combinations of rules and settings with no luck.

**Comment by bradzacher** [comment]
> EXPERIMENTAL_useSourceOfProjectReferenceRedirect: true

I would avoid using this option because it's half-baked and not properly built.
There are many issues with it which is why we haven't progressed on it.
Does turning it off fix the issue?

**Comment by milesj** [comment]
It seems to work for some projects but not others. I also noticed that `import/export` caused issues, and disabling that fixed some more projects. But I still can't get this to work 100%, and I'm not seeing any patterns.

**Comment by bradzacher** [comment]
If you can isolate a repro for that I'd be more than happy to dig deeper and help out!

**Comment by JoshuaKGoldberg** [comment]
Several months of no activity, awaiting response + unable to repro. I'm going to go ahead and close this out as non-actionable. Please let me know if you do have a repro for us to look at @milesj - and thanks for working with us on trying to narrow it down! ❤️

**Comment by milesj** [comment]
That's fine. I was never able to get this working nor find a root cause.

**Comment by aaronadamsCA** [comment]
@JoshuaKGoldberg Hey, I've got a repro of what I'm pretty sure is the same issue.

https://github.com/aaronadamsCA/typescript-eslint-issue-5136

I seem to have it narrowed down to the following _very specific_ combination:

- Using `typescript-eslint` together with `eslint-plugin-import`;
- Using `allowAutomaticSingleRunInference`;
- Importing from `"@mui/joy/styles"` - it seems this import is haunted somehow.

Even weirder - it doesn't always reproduce locally, but fails every time when running on GitHub Actions:

https://github.com/aaronadamsCA/typescript-eslint-issue-5136/actions/runs/3321426085/jobs/5489056308

This was not easy to isolate, but hopefully this reproduction can help resolve the issue for good. For now, we will simply disable `allowAutomaticSingleRunInference`, which works around the issue.

**Comment by JoshuaKGoldberg** [comment]
Ah fantastic, thanks for posting @aaronadamsCA! Could you do me a favor and file a new issue with it please? This one's got a long thread and I want to make sure this info doesn't get buried.
